parser/parser_test.go

```diff
@@ -639,23 +639,276 @@ func TestParseValidInput(t *testing.T) {
+	default:
+		return comparePrimitives(actual, expected, depth)
+	}
+
+}
+
+func deepComparePointers(actual namedItem, expected namedItem, depth int) (equal bool, diff string) {
+	if actual.item.IsNil() || expected.item.IsNil() {
+		if actual.item.IsNil() && expected.item.IsNil() {
+			return true, ""
+		} else {
+
+			if actual.item.IsNil() {
+				return false, fmt.Sprintf("%s is nil whereas %s is %#v", actual.name, expected.name, expected.item)
+			} else {
+				return false, fmt.Sprintf("%s is nil whereas %s is %#v", expected.name, actual.name, actual.item)
 			}
+		}
+
+	}
+	return deepCompareObjects(namedItem{actual.name, actual.item.Elem()},
+		namedItem{expected.name, expected.item.Elem()},
+		depth)
+}
+
+func comparePrimitives(actual namedItem, expected namedItem, depth int) (equal bool, diff string) {
+	// TODO make this less hacky
+	var a, b interface{}
+	if actual.item.Kind() != expected.item.Kind() {
+		return false, fmt.Sprintf("%s is of type %s whereas %s is of type %s", actual.name, actual.item.Kind(), expected.name, expected.item.Kind())
+	}
+	switch actual.item.Kind() {
+	case reflect.String:
+		a = actual.item.String()
+		b = expected.item.String()
+	case reflect.Bool:
+		a = actual.item.Bool()
+		b = expected.item.Bool()
+	case reflect.Int:
+		a = actual.item.Int()
+		b = expected.item.Int()
+	default:
+		panic(fmt.Sprintf("unrecognized object types, %s (%#v) and %s (%#v)", actual.name, actual.item, expected.name, expected.item))
+	}
+	if reflect.DeepEqual(a, b) {
+		return true, ""
+	} else {
+		return false, fmt.Sprintf("%s = %#v whereas %s = %#v", actual.name, actual.item, expected.name, expected.item)
+	}
+
+}
+
+func deepCompareMaps(actual namedItem, expected namedItem, depth int) (equal bool, diff string) {
+	a := actual.item
+	b := expected.item
+	aKeys := a.MapKeys()
+	bKeys := b.MapKeys()
+	for _, aKey := range aKeys {
+		if !(isIn(aKey, bKeys)) {
+			aValue := a.MapIndex(aKey)
+			return false, fmt.Sprintf("%v contains key %s (with value %#v) but %v does not", actual.name, aKey, aValue, expected.name)
+		}
+	}
+	for _, bKey := range bKeys {
+		if !(isIn(bKey, aKeys)) {
+			bValue := b.MapIndex(bKey)
+			return false, fmt.Sprintf("%v contains key %s (with value %#v) but %v does not", expected.name, bKey, bValue, actual.name)
+		}
+	}
+	for _, aKey := range aKeys {
+		bKey, _ := findMatching(aKey, bKeys)
+		aIndexText := fmt.Sprintf("[%v(%t)]", aKey, bKey)
+		bIndexText := fmt.Sprintf("[%v(%t)]", bKey, bKey)
+
+		equal, diff = deepCompareObjects(namedItem{actual.name + aIndexText, a.MapIndex(aKey)},
+			namedItem{expected.name + bIndexText, b.MapIndex(bKey)},
+			depth,
+		)
+		if !equal {
+			return equal, diff
+		}
+
+	}
+	return true, ""
+}
+
+func deepCompareArrays(actual namedItem, expected namedItem, depth int) (equal bool, diff string) {
+	a := actual.item
+	b := expected.item
+	aLen := a.Len()
+	bLen := b.Len()
+	sharedLen := int(math.Min(float64(aLen), float64(bLen)))
+	for i := 0; i < sharedLen; i++ {
```

Comment:
I'll re-implement the 'min' function as you mentioned above.

I extracted the 'min' operation to slightly decrease the number of values being recomputed during each loop iteration (I don't see anything in https://github.com/golang/go/wiki/CompilerOptimizations specifying that the Go compiler can optimize 'i < a.Len() && i < b.Len()').